AX: List markers should be stitched into adjacent text in the accessibility tree

twilco opened the pull request:
#### bcf7cbeeebf02dc79ab17e0aed58ecf87e822c62
<pre>
AX: List markers should be stitched into adjacent text in the accessibility tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=303552">https://bugs.webkit.org/show_bug.cgi?id=303552</a>
<a href="https://rdar.apple.com/165841939">rdar://165841939</a>

Reviewed by Joshua Hoffman.

One common complaint is that list markers are an extra navigation stop in many assistive technologies like VoiceOver.
This commit fixes that issue by stitching the list marker with adjacent text.

Mechanically, this required some refactoring, since the prior implementation of stitching assumed that the first
member of a stitch group was always the representative in the accessibility tree, and that it had a node. This
is not true when considering list markers, which are generally first in the inline layout tree (which is what we
walk to build stitch groups), and do not have an associated node.

Fix this by creating a new AXStitchGroup class that denotes the ID of the group&apos;s representative, and handle the case
where element&apos;s behind the representative may have text or geometry to contribute to the serving of our accessibility APIs.

Test: accessibility/list-marker-text-stitching.html

* LayoutTests/accessibility/list-marker-text-stitching-expected.txt: Added.
* LayoutTests/accessibility/list-marker-text-stitching.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new test.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/list-marker-text-stitching-expected.txt: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::stitchGroupFromGroups const):
(WebCore::AXCoreObject::stitchStateFromGroups const): Deleted.
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::hasStitchableRole const):
(WebCore::AXCoreObject::stitchGroup const):
(WebCore::AXCoreObject::stitchedIntoID const):
(WebCore::AXCoreObject::StitchState::StitchState): Deleted.
(WebCore::AXCoreObject::stitchState const): Deleted.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::stitchGroupsOwnedBy):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXStitchGroup.h: Copied from Source/WebCore/accessibility/AXStitchUtilities.h.
(WebCore::AXStitchGroup::AXStitchGroup):
(WebCore::AXStitchGroup::isEmpty const):
(WebCore::AXStitchGroup::isValid const):
(WebCore::AXStitchGroup::members const):
(WebCore::AXStitchGroup::representativeID const):
* Source/WebCore/accessibility/AXStitchUtilities.cpp:
* Source/WebCore/accessibility/AXStitchUtilities.h:
* Source/WebCore/accessibility/AXUtilities.cpp:
(WebCore::lastNode):
(WebCore::lastObject): Deleted.
* Source/WebCore/accessibility/AXUtilities.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::stitchGroups const):
(WebCore::AccessibilityNodeObject::stitchGroup const):
(WebCore::AccessibilityNodeObject::stringValue const):
(WebCore::AccessibilityNodeObject::stitchState const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::stitchGroups const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::boundingBoxRect const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::isDefaultValue):
(WebCore::AXIsolatedObject::stitchGroupsView const):
(WebCore::AXIsolatedObject::stitchGroup const):
(WebCore::AXIsolatedObject::relativeFrame const):
(WebCore::AXIsolatedObject::stringValue const):
(WebCore::AXIsolatedObject::stitchState const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/rendering/RenderBlock.h:

Canonical link: <a href="https://commits.webkit.org/303962@main">https://commits.webkit.org/303962@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80246607ffb0bc771928b48199381f9bd20ae308

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134048 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6559 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45253 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141628 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86110 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135918 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6423 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102528 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/badbaa74-01e0-4f4f-a676-e17c1c115539) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136995 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4968 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120173 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83326 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3c86ffda-16f5-4fab-887f-eb40cdc35b0c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4846 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2475 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1443 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114018 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38294 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144274 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6229 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38871 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110900 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6311 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5271 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111118 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28191 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4711 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116430 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59999 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6281 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34671 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6127 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69745 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6372 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6235 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -268,6 +268,16 @@ function expect(expression, expectedValue) {
     return `FAIL: ${expression} !== ${expectedValue}, was ${eval(expression)}\n`;
 }
 
+function expectNumber(expression, expectedValue, allowedVariance = 0) {
```
Thanks for adding this! This will be super helpful down the road

minorninth commented on Source/WebCore/accessibility/AXStitchGroup.h:
```diff
@@ -0,0 +1,68 @@
+
+namespace WebCore {
+
+class AXStitchGroup {
```
Add a class comment? Explain what this represents, who owns it, when it's valid, etc.

twilco replied:
Comment added!

twilco committed:
AX: List markers should be stitched into adjacent text in the accessibility tree
https://bugs.webkit.org/show_bug.cgi?id=303552
rdar://165841939

Reviewed by Joshua Hoffman.

One common complaint is that list markers are an extra navigation stop in many assistive technologies like VoiceOver.
This commit fixes that issue by stitching the list marker with adjacent text.

Mechanically, this required some refactoring, since the prior implementation of stitching assumed that the first
member of a stitch group was always the representative in the accessibility tree, and that it had a node. This
is not true when considering list markers, which are generally first in the inline layout tree (which is what we
walk to build stitch groups), and do not have an associated node.

Fix this by creating a new AXStitchGroup class that denotes the ID of the group's representative, and handle the case
where element's behind the representative may have text or geometry to contribute to the serving of our accessibility APIs.

Test: accessibility/list-marker-text-stitching.html

* LayoutTests/accessibility/list-marker-text-stitching-expected.txt: Added.
* LayoutTests/accessibility/list-marker-text-stitching.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new test.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/list-marker-text-stitching-expected.txt: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::stitchGroupFromGroups const):
(WebCore::AXCoreObject::stitchStateFromGroups const): Deleted.
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::hasStitchableRole const):
(WebCore::AXCoreObject::stitchGroup const):
(WebCore::AXCoreObject::stitchedIntoID const):
(WebCore::AXCoreObject::StitchState::StitchState): Deleted.
(WebCore::AXCoreObject::stitchState const): Deleted.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::stitchGroupsOwnedBy):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXStitchGroup.h: Copied from Source/WebCore/accessibility/AXStitchUtilities.h.
(WebCore::AXStitchGroup::AXStitchGroup):
(WebCore::AXStitchGroup::isEmpty const):
(WebCore::AXStitchGroup::isValid const):
(WebCore::AXStitchGroup::members const):
(WebCore::AXStitchGroup::representativeID const):
* Source/WebCore/accessibility/AXStitchUtilities.cpp:
* Source/WebCore/accessibility/AXStitchUtilities.h:
* Source/WebCore/accessibility/AXUtilities.cpp:
(WebCore::lastNode):
(WebCore::lastObject): Deleted.
* Source/WebCore/accessibility/AXUtilities.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::stitchGroups const):
(WebCore::AccessibilityNodeObject::stitchGroup const):
(WebCore::AccessibilityNodeObject::stringValue const):
(WebCore::AccessibilityNodeObject::stitchState const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::stitchGroups const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::boundingBoxRect const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::isDefaultValue):
(WebCore::AXIsolatedObject::stitchGroupsView const):
(WebCore::AXIsolatedObject::stitchGroup const):
(WebCore::AXIsolatedObject::relativeFrame const):
(WebCore::AXIsolatedObject::stringValue const):
(WebCore::AXIsolatedObject::stitchState const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/rendering/RenderBlock.h:

Canonical link: https://commits.webkit.org/303962@main
